File: verletplus/src/objects/verlet/object.cpp
#include <iostream>
#include "../../physics/vector2d.cpp"
#include "../../utils/colors.cpp"
#include "../../utils/time.cpp"

class VerletObject
{
public:
    Vector2D current_position;
    Vector2D prev_position;
    Vector2D acceleration;
    Vector2D velocity;
    Color color = Colors().WHITE;
    float start_time;

    VerletObject(Vector2D position, Color color) {
        this->current_position = position;
        this->prev_position = position;
        this->acceleration = Vector2D();
        this->velocity = Vector2D();
        this->color = color;
        this->start_time = Time().seconds();
    }

    // Set the object's color
    void set_color(Color color) {
        this->color = color;
    }

    // Calculate the object's velocity
    Vector2D calculate_velocity() {
        return this->current_position - this->prev_position;
    }

    // Perform the verlet integration to calculate the displacement
    Vector2D calculate_displacement(float dt) {
        return this->current_position + this->velocity + this->acceleration * dt;
    }

    // Accelerate the object
    void accelerate(Vector2D acceleration) {
        this->acceleration += acceleration;
    }

    // Update the object's position
    void update_position() {
        // Get the time difference (delta time)
        float dt = Time().seconds() - this->start_time;

        // Update the object's position
        this->velocity = this->calculate_velocity();
        this->prev_position = this->current_position.copy();
        this->current_position = this->calculate_displacement(dt);

        // Reset the acceleration and start time
        this->acceleration.zero();
        this->start_time = Time().seconds();
    }
};

File: python/objects/verlet/object.py
from physics import Vector2D
import time


class VerletObject(object):
    def __init__(
        self,
        position: tuple[float, float],
        color: tuple[int, int, int] = (255, 255, 255),
    ) -> None:
        self.current_position: Vector2D = Vector2D(position[0], position[1])
        self.prev_position: Vector2D = Vector2D(position[0], position[1])
        self.acceleration: Vector2D = Vector2D(0.0, 0.0)
        self.velocity: Vector2D = Vector2D(0.0, 0.0)
        self.color: tuple[int, int, int] = color
        self.start_time: float = time.time()

    # Update the ball's color
    def set_color(self, color: tuple[int, int, int]) -> None:
        self.color = color

    # Calculate the objects velocity
    def calculate_velocity(self) -> Vector2D:
        return self.current_position - self.prev_position

    # Perform the verlet integration to calcualte the displacement
    def calculate_displacement(self, dt: float):
        return self.current_position + self.velocity + self.acceleration * dt

    # Accelerate the object
    def accelerate(self, acceleration: Vector2D) -> None:
        self.acceleration += acceleration

    # Update the objects position
    def update_position(self) -> None:
        dt: float = time.time() - self.start_time
        self.velocity = self.calculate_velocity()
        self.prev_position = self.current_position.copy()
        self.current_position = self.calculate_displacement(dt)
        self.acceleration.zero()
        self.start_time = time.time()
